Name: Swapnanil Halder

Roll No. : 18MA20046

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler as scale

### Data Importion

In [2]:
data = pd.read_csv(r"./data.csv", header=0)
data.tail(3)

age  sex  cp trestbps chol fbs restecg thalach exang  oldpeak slope ca  \
291   56    1   4      155  342   1       0     150     1      3.0     2  ?   
292   58    0   2      180  393   0       0     110     1      1.0     2  ?   
293   65    1   4      130  275   0       1     115     1      1.0     2  ?   

    thal  num         
291    ?           1  
292    7           1  
293    ?           1

In [3]:
data.head(3)

age  sex  cp trestbps chol fbs restecg thalach exang  oldpeak slope ca  \
0   28    1   2      130  132   0       2     185     0      0.0     ?  ?   
1   29    1   2      120  243   0       0     160     0      0.0     ?  ?   
2   29    1   2      140    ?   0       0     170     0      0.0     ?  ?   

  thal  num         
0    ?           0  
1    ?           0  
2    ?           0

In [4]:
data.drop(labels=['slope', 'ca', 'thal'], axis=1, inplace=True)

In [5]:
data.head(3)

age  sex  cp trestbps chol fbs restecg thalach exang  oldpeak  num       
0   28    1   2      130  132   0       2     185     0      0.0           0
1   29    1   2      120  243   0       0     160     0      0.0           0
2   29    1   2      140    ?   0       0     170     0      0.0           0

In [6]:
## To drop rows containing missing values
data.replace('?', np.nan, inplace=True)
data.dropna(axis=0, how='any', inplace=True)
data.reset_index(drop=True, inplace=True)

In [7]:
print(data.shape)
data.head(3)

(261, 11)


age  sex  cp trestbps chol fbs restecg thalach exang  oldpeak  num       
0   28    1   2      130  132   0       2     185     0      0.0           0
1   29    1   2      120  243   0       0     160     0      0.0           0
2   30    0   1      170  237   0       1     170     0      0.0           0

In [8]:
print(data.columns)

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'num       '],
      dtype='object')


In [9]:
data_X, data_Y = data.loc[:, ~data.columns.isin(['num       '])], data.loc[:, ['num       ']]
X, Y = data_X.to_numpy(), data_Y.to_numpy().reshape(-1)
X[:,:] = scale().fit_transform(X[:,:])

In [10]:
Y[:]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

## Linear Regression

In [11]:
class LinearRegression() :
    
    def __init__( self, learning_rate, iterations ) :
        self.learning_rate = learning_rate
        self.iterations = iterations

        
    # Function for model training
    def fit( self, X, Y ) :

        # no_of_training_examples, no_of_features
        self.m, self.n = X.shape

        # weight initialization

        self.W = np.zeros( self.n )
        self.b = 0

        self.X = X
        self.Y = Y

        # gradient descent learning
        for i in range( self.iterations ) :
            self.update_weights()

        return self

    
    # Helper function to update weights in gradient descent
    def update_weights( self ) :

        Y_pred = self.predict( self.X )

        # calculate gradients
        dW =  -( 2 * ( self.X.T ).dot( self.Y - Y_pred ) )
        db =  -2 * np.sum( self.Y - Y_pred )

        # update weights
        self.W = self.W - self.learning_rate * dW
        self.b = self.b - self.learning_rate * db

        return self


    def predict( self, X ) :
        return X.dot( self.W ) + self.b
    
    
    # Mean square error for testing data
    def mse( self, test, pred ):
        temp = test - pred
        return np.dot(temp, temp.T) / len(test)
    
    
    # Calculating accuracy with given treshold
    def accuracyy( self, test, pred, treshold):
        
        yy = []
        for i in pred:
            yy.append(1 if i>=treshold else 0)
            
        count = 0
        for i in range(len(yy)):
            if yy[i] == test[i]:
                count += 1
                
        return (count*1.0) / (len(yy))


In [12]:

# Splitting dataset into train and test set
X_train, X_test, Y_train, Y_test = train_test_split( X, Y, test_size = 1/3, random_state = 0 )


# Model training
model = LinearRegression( iterations = 1000, learning_rate = 0.001 )
model.fit( X_train, Y_train )

In [13]:
# Prediction on test set
Y_pred = model.predict( X_test )
Y_pred[:5]

array([0.17736460543539645, -0.03783453348329134, 0.9047983788358998,
       0.08464377899895091, 0.9207529500001439], dtype=object)

In [14]:
Y_test[:5]

array([1, 0, 1, 0, 1])

In [15]:
print("Mean square error using Linear regression model : ", model.mse(Y_test, Y_pred))

Mean square error using Linear regression model :  0.12164722625804501


Here We have predicted the values for test data but we need to classify them as 0 or 1 so, we use a treshold of 0.5.                                              
If the vlaue is greater than 0.5 then we make it 1 else 0

In [16]:
print("Accuracy of Linear Regression is : ", np.round(model.accuracyy(Y_test , Y_pred, 0.5)*100, 2), "%")

Accuracy of Linear Regression is :  86.21 %


***

## Ridge Regression

In [17]:
class RidgeRegression() :

    def __init__( self, learning_rate, iterations, l2_penality ) :

        self.learning_rate = learning_rate
        self.iterations = iterations
        self.l2_penality = l2_penality

    # Function for model training
    def fit( self, X, Y ) :

        # no_of_training_examples, no_of_features
        self.m, self.n = X.shape

        # weight initialization
        self.W = np.zeros( self.n )
        self.b = 0
        
        self.X = X
        self.Y = Y

        # gradient descent learning
        for i in range( self.iterations ) :
            self.update_weights()
        return self

    

    def update_weights( self ) :
        Y_pred = self.predict( self.X )

        # calculate gradients
        dW = ( - ( 2 * ( self.X.T ).dot( self.Y - Y_pred ) ) + ( 2 * self.l2_penality * self.W ) ) 
        db = - 2 * np.sum( self.Y - Y_pred ) 

        # update weights
        self.W = self.W - self.learning_rate * dW
        self.b = self.b - self.learning_rate * db
        return self


    def predict( self, X ) :
        return X.dot( self.W ) + self.b
    
    
    # Mean square error for testing data
    def mse( self, test, pred ):
        temp = test - pred
        return np.dot(temp, temp.T) / len(test)

    
    # Calculating accuracy with given treshold
    def accuracyy( self, test, pred, treshold):
        
        yy = []
        for i in pred:
            yy.append(1 if i>=treshold else 0)
            
        count = 0
        for i in range(len(yy)):
            if yy[i] == test[i]:
                count += 1
                
        return (count*1.0) / (len(yy))

In [18]:
def find_lambda_Ridge( X, Y ):
    
    # Splitting dataset into train and test set
    X_train, X_test, Y_train, Y_test = train_test_split( X, Y, test_size = 1 / 3, random_state = 0 )

    max_accu = 0
    min_lambda = 1
    
    for lambdaa in range(100):
        # Model training
        model = RidgeRegression( iterations = 1000, learning_rate = 0.001, l2_penality = lambdaa )
        model.fit( X_train, Y_train )
    
        Y_pred = model.predict( X_test )

        temp = model.accuracyy( Y_test, Y_pred, 0.5 )

        if temp > max_accu:
            min_lambda = lambdaa
            max_accu = temp
        
    return min_lambda

In [19]:
## Now we calculate for which lambda the accuracy is highest for Ridge regression

min_lam = find_lambda_Ridge( X, Y )
min_lam 

0

In [20]:
# Splitting dataset into train and test set
X_train, X_test, Y_train, Y_test = train_test_split( X, Y, test_size = 1 / 3, random_state = 0 )

# Model training
model = RidgeRegression( iterations = 1000, learning_rate = 0.001, l2_penality = min_lam )
model.fit( X_train, Y_train )

In [21]:
Y_pred = model.predict( X_test )
Y_pred[:5]

array([0.17736460543539645, -0.03783453348329134, 0.9047983788358998,
       0.08464377899895091, 0.9207529500001439], dtype=object)

In [22]:
Y_test[:5]

array([1, 0, 1, 0, 1])

In [23]:
print("Mean square error using Ridge regression model : ", model.mse( Y_test, Y_pred ))

Mean square error using Ridge regression model :  0.12164722625804501


In [24]:
print("Accuracy of Ridge Regression is : ", np.round(model.accuracyy(Y_test , Y_pred, 0.5)*100, 2), "%")

Accuracy of Ridge Regression is :  86.21 %


*****************************************************************

## Lasso Regression 

In [25]:
class LassoRegression() :

    def __init__( self, learning_rate, iterations, l1_penality ) :

        self.learning_rate = learning_rate
        self.iterations = iterations
        self.l1_penality = l1_penality

    # Function for model training
    def fit( self, X, Y ) :

        # no_of_training_examples, no_of_features
        self.m, self.n = X.shape

        # weight initialization
        self.W = np.zeros( self.n )
        self.b = 0
        
        self.X = X
        self.Y = Y

        # gradient descent learning

        for i in range( self.iterations ) :
            self.update_weights()
        return self

    # Helper function to update weights in gradient descent

    def update_weights( self ) :
        Y_pred = self.predict( self.X )

        # calculate gradients
        sign = []
        for val in self.W:
            sign.append(1 if val > 0 else -1)
        sign = np.array(sign)
        
        dW = ( - ( 2 * ( self.X.T ).dot( self.Y - Y_pred ) ) + ( self.l1_penality * sign ) ) 
        db = - 2 * np.sum( self.Y - Y_pred ) 

        # update weights
        self.W = self.W - self.learning_rate * dW
        self.b = self.b - self.learning_rate * db
        return self

    # Hypothetical function h( x )c
    def predict( self, X ) :
        return X.dot( self.W ) + self.b
    
    
    # Mean square error for testing data
    def mse( self, test, pred ):
        temp = test - pred
        return np.dot(temp, temp.T) / len(test)
    
    # Calculating accuracy with given treshold
    def accuracyy( self, test, pred, treshold):
        
        yy = []
        for i in pred:
            yy.append(1 if i>=treshold else 0)
            
        count = 0
        for i in range(len(yy)):
            if yy[i] == test[i]:
                count += 1
                
        return (count*1.0) / (len(yy))

In [26]:
def find_lambda_Lasso( X, Y ):
    
    # Splitting dataset into train and test set
    X_train, X_test, Y_train, Y_test = train_test_split( X, Y, test_size = 1 / 3, random_state = 0 )

    max_accu = 0
    min_lambda = 1
    
    for lambdaa in range(100):
        # Model training
        model = LassoRegression( iterations = 1000, learning_rate = 0.001, l1_penality = lambdaa )
        model.fit( X_train, Y_train )
    
        Y_pred = model.predict( X_test )
        
        temp = model.accuracyy( Y_test, Y_pred, 0.5 )

        if temp > max_accu:
            min_lambda = lambdaa
            max_accu = temp
    
        
    return min_lambda

In [27]:
## Now we calculate for which lambda the accuracy is highest for Lasso regression

min_lam = find_lambda_Lasso(X, Y)
min_lam

0

In [28]:
# Splitting dataset into train and test set
X_train, X_test, Y_train, Y_test = train_test_split( X, Y, test_size = 1 / 3, random_state = 0 )

# Model training
model = LassoRegression( iterations = 1000, learning_rate = 0.001, l1_penality = min_lam )
model.fit( X_train, Y_train )

In [29]:
Y_pred = model.predict( X_test )
Y_pred[:5]

array([0.17736460543539645, -0.03783453348329134, 0.9047983788358998,
       0.08464377899895091, 0.9207529500001439], dtype=object)

In [30]:
Y_test[:5]

array([1, 0, 1, 0, 1])

In [31]:
print("Mean square error using Lasso regression model : ", model.mse( Y_test, Y_pred ))

Mean square error using Lasso regression model :  0.12164722625804501


In [32]:
print("Accuracy of Lasso Regression is : ", np.round(model.accuracyy(Y_test , Y_pred, 0.5)*100, 2), "%")

Accuracy of Lasso Regression is :  86.21 %


*****************************************************************

## Using Simplest Model (Rough Model)

In [33]:
def mse( test, pred ):
        temp = test - pred
        return np.dot(temp, temp.T) / len(test)
    

In [34]:
def accuracyy( test, pred, treshold ):
        
        yy = []
        for i in pred:
            yy.append(1 if i>=treshold else 0)
            
        count = 0
        for i in range(len(yy)):
            if yy[i] == test[i]:
                count += 1
                
        return (count*1.0) / (len(yy))

In [35]:
np.mean(Y_train)  

0.3850574712643678

In [36]:
## this simplest model always gives mean of Y_train for every X_test 
    
pred_val = np.mean(Y_train) 
 
pred = [pred_val*1 for i in range(len(Y_test))]
pred = np.array(pred)

In [37]:
print("Mean square error using Simplest model : ", mse( Y_test, pred ))

Mean square error using Simplest model :  0.2301823226317875


In [38]:
print("Accuracy of Simplest Model is : ", np.round(accuracyy( Y_test , pred, 0.5 )*100, 2), "%")

Accuracy of Simplest Model is :  64.37 %


## Conclusion

As we have seen that all regression models ( linear, Ridge, Lasso ) has given the same accuracy ( 86.21 %).

This may happen as there is very less data